sklearn/utils/_metadata_requests.py

```diff
@@ -368,13 +378,46 @@ class MetadataRequest:
     def __init__(self, owner):
         # this is used to check if the user has set any request values
         self._is_default_request = False
-        for method in METHODS:
+        self.owner = owner
+        for method in SIMPLE_METHODS:
             setattr(
                 self,
                 method,
                 MethodMetadataRequest(router=self, owner=owner, method=method),
             )
 
+    def __getattr__(self, name):
+        # Called when the default attribute access fails with an AttributeError
+        # (either __getattribute__() raises an AttributeError because name is
+        # not an instance attribute or an attribute in the class tree for self;
+        # or __get__() of a name property raises AttributeError). This method
+        # should either return the (computed) attribute value or raise an
+        # AttributeError exception.
+        if name not in COMPOSITE_METHODS:
+            raise AttributeError(
+                f"'{self.__class__.__name__}' object has no attribute '{name}'"
+            )
+
+        res = MethodMetadataRequest(router=self, owner=self.owner, method=name)
+        res._requests = {}
+        for method in COMPOSITE_METHODS[name]:
+            mmr = getattr(self, method)
+            existing = set(res._requests.keys())
+            upcoming = set(mmr._requests.keys())
+            common = existing & upcoming
+            conflicts = [
+                key for key in common if res._requests[key] != mmr._requests[key]
+            ]
+            if conflicts:
+                raise ValueError(
+                    f"Conflicting metadata requests for {', '.join(conflicts)} while"
+                    f" composing the requests for {name}. Metadata with the same name"
+                    f" for methods {', '.join(COMPOSITE_METHODS[name])} should have the"
+                    " same request value."
+                )
+            res._requests.update(mmr._requests)
```

Comment:
Having `MetadataRequest` interactive directly with `_requests` feels strange from a OOP design point of view.

Should `MethodMetadataRequest` allow for `request` to be passed in, so that `MetadataRequest` can construct the request dict and then passing it into `MethodMetadataRequest.__init__`?

Reply:
If I understand properly, @thomasjpfan you would like the following pattern:

```python
        requests = {}
        for method in COMPOSITE_METHODS[name]:
            mmr = getattr(self, method)
            existing = set(requests.keys())
            upcoming = set(mmr._requests.keys())
            common = existing & upcoming
            conflicts = [
                key for key in common if requests[key] != mmr._requests[key]
            ]
            if conflicts:
                raise ValueError(
                    f"Conflicting metadata requests for {', '.join(conflicts)} while"
                    f" composing the requests for {name}. Metadata with the same name"
                    f" for methods {', '.join(COMPOSITE_METHODS[name])} should have the"
                    " same request value."
                )
            requests.update(mmr._requests)
        return MethodMetadataRequest(
            router=self, owner=self.owner, method=name, request=request
        )
```

Reply:
If this is the case, it is true is a bit less strange than updating interactively the private attribute.

Reply:
Yes, https://github.com/scikit-learn/scikit-learn/pull/24585/files#r1012951287 is exactly I had in mind.

Reply:
Applied the suggestion.